contracts/interfaces/New.sol

```diff
@@ -0,0 +1,135 @@
+        require(_identityId != bytes32(0), "invalid identity id"); // is it the case that identity id will never be 0? any reason it'd be better to use pool?
+        require(identityId == bytes32(0), "already initialized");
+        identityId = _identityId;
+        poolId = _poolId;
+
+        initializeRecipientEligibilityModule(_recipientEligibilityData);
+        initializeVoterEligibilityModule(_voterEligibilityData);
+        initializeVotingModule(_votingData);
+        initializeAllocationModule(_allocationData);
+        initializeDistributionModule(_distributionData);
+    }
+
+    function poolFunded(uint256 _amount) external;
+    // plus maybe some modifiers
+}
+
+interface IRecipientEligibilityModule {
+    enum RecipientStatus {
+        None,
+        Pending,
+        Accepted,
+        Rejected
+    }
+
+    function initializeRecipientEligibilityModule(bytes _data) external;
+    function registerRecipients(bytes memory _data, address _sender) external payable returns (address);
+    function getRecipientStatus(address _recipientId) public view returns (RecipientStatus);
+    function allRecipients() public view returns (address[] memory);
+}
+
+interface IVoterEligibilityModule {
+    function initializeVoterEligibilityModule(bytes _data) external;
+    function isValidVoter(address _voter) public view returns (bool); // maybe should be renamed allocator, unless we change to vote below?
+}
+
+interface IVotingModule {
+    function initializeVotingModule(bytes _data) external;
+    function vote(bytes memory _data, address _sender) external payable; //  is this the right name given module changes?
+    function getPayout(address[] memory _recipientId, bytes memory _data)
+        external
+        view
+        returns (PayoutSummary[] memory summaries);
+    // maybe readyToPayout? although i think getPayout should just return 0? not sure yet.
+}
+
+interface IAllocationModule {
+    function initializeAllocationModule(bytes _data) external;
+}
+
+interface IDistributionModule {
+    function initializeDistributionModule(bytes _data) external;
+    function distribute(address[] memory _recipientIds, bytes memory _data, address _sender) external;
+}
+
+contract WinnerTakesAll {
+    mapping(address => uint256) public votes;
+
+    function allocate(address recipient) external payable {
+        votes[recipient] += 1;
+    }
+
+    function getResults(address[] memory recipientIds) public view returns (uint256[]) {
+        uint256[] memory results = new uint[](recipientIds.length);
+        for (uint256 i = 0; i < recipientIds.length; i++) {
+            results[i] = votes[recipientIds[i]];
+        }
+        return results;
+    }
+
+    function getPayouts(address[] memory recipientIds) public view returns (PayoutSummary[]) {
+        require(recipientIds.length == 0, "pass empty rec, will use all");
+        address[] memory recipients = allRecipients();
+        uint256[] memory results = getResults(recipients);
```

Comment:
I think we shouldn't rely that heavily on arrays.
imagine having an `allReceipients` array with thousands of addresses. it would be very gas consuming and maybe end up in a dos condition by reaching the block gas limit.

I think we need a better way to wire the modules. i could imagine using internal functions as hooks for every state modifying function  between the modules.

```
function allocate(address recipient) external payable {
        votes[recipient] += 1;
    }
 ```
 would become to
 ```
 function allocate(address recipient) external payable {
        votes[recipient] += 1;
       _allocate(address recipient);
    }
 ```
 And the `WinnerTakesAll` module would implement something like:
 
 ```
 address maxAllocationAddress;

function _allocate(address _receipient) internal {
   if(votes[_receipient] > votes[maxAllocationAddress]) {
        maxAllocationAddress = _receipient;
   }
}
// does it makes sense to split it into an internal and a external function instead of a public one?
function getPayouts(address[] memory recipientIds) public view returns (PayoutSummary[]) {
        require(recipientIds.length == 0, "pass empty rec, will use all");

       PayoutSummary[] memory payouts = new PayoutSummary[](1);
        for (uint256 i = 0; i < recipientIds.length; i++) {
            if (recipients[i] == maxAllocationAddress) {
                payouts[0] = PayoutSummary(recipients[i], address(this).balance);
            }
        }
        return payouts;
```

Tbh i'm currently not sure how we would implement the internal hook if multiple modules would need to implement the hook. but it feels like exploring this path could save a lot of heavy calculations and gas.

@zobront @thelostone-mc @codenamejason